File: frontend/src/services/roadmap.ts
import { api } from './api';

export interface RoadmapTask {
  id: number;
  phase_id: number;
  title: string;
  description?: string;
  resource_links?: string;
  practical_exercise?: string;
  interview_tips?: string;
  is_completed: boolean;
}

export interface RoadmapPhase {
  id: number;
  roadmap_id: number;
  title: string;
  description?: string;
  estimated_duration?: string;
  phase_order: number;
  is_completed: boolean;
  tasks: RoadmapTask[];
}

export interface Roadmap {
  id: number;
  user_id: number;
  target_role: string;
  current_status?: string;
  created_at: string;
  phases: RoadmapPhase[];
}

export const generateRoadmap = async (targetRole: string) => {
  return await api.post('/roadmaps', { target_role: targetRole });
};

export const getAllRoadmaps = async () => {
  return await api.get('/roadmaps');
};

export const getRoadmapById = async (roadmapId: string | number) => {
  return await api.get(`/roadmaps/${roadmapId}`);
};

export const updateTaskStatus = async (taskId: number, isCompleted: boolean) => {
  return await api.patch(`/roadmaps/tasks/${taskId}`, { is_completed: isCompleted });
};

export const deleteRoadmap = async (roadmapId: string | number) => {
  return await api.delete(`/roadmaps/${roadmapId}`);
};

File: backend/models/roadmap.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Text, Boolean
from sqlalchemy.orm import relationship
from db.database import Base
from sqlalchemy.sql import func

class Roadmap(Base):
    __tablename__ = "roadmaps"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    target_role = Column(String, nullable=False)
    current_status = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    phases = relationship("RoadmapPhase", back_populates="roadmap", cascade="all, delete-orphan")

class RoadmapPhase(Base):
    __tablename__ = "roadmap_phases"
    
    id = Column(Integer, primary_key=True, index=True)
    roadmap_id = Column(Integer, ForeignKey("roadmaps.id", ondelete="CASCADE"), nullable=False)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    estimated_duration = Column(String, nullable=True) # NEW FIELD
    phase_order = Column(Integer, nullable=False)
    is_completed = Column(Boolean, default=False)
    
    roadmap = relationship("Roadmap", back_populates="phases")
    tasks = relationship("RoadmapTask", back_populates="phase", cascade="all, delete-orphan")

class RoadmapTask(Base):
    __tablename__ = "roadmap_tasks"
    
    id = Column(Integer, primary_key=True, index=True)
    phase_id = Column(Integer, ForeignKey("roadmap_phases.id", ondelete="CASCADE"), nullable=False)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    resource_links = Column(Text, nullable=True)
    practical_exercise = Column(Text, nullable=True) # NEW FIELD
    interview_tips = Column(Text, nullable=True)     # NEW FIELD
    is_completed = Column(Boolean, default=False)
    
    phase = relationship("RoadmapPhase", back_populates="tasks")